pkg/distributor/distributor.go

```diff
@@ -100,52 +101,52 @@ type Distributor struct {
+	ingestionRateLimiter *limiter.RateLimiter
 }
 
 // Config contains the configuration require to
 // create a Distributor
 type Config struct {
-	EnableBilling bool                       `yaml:"enable_billing,omitempty"`
-	BillingConfig billing.Config             `yaml:"billing,omitempty"`
-	PoolConfig    ingester_client.PoolConfig `yaml:"pool,omitempty"`
+	EnableBilling bool                                      `yaml:"enable_billing,omitempty"`
+	BillingConfig billing.Config                            `yaml:"billing,omitempty"`
+	PoolConfig    ingester_client.PoolConfig                `yaml:"pool,omitempty"`
+	Registry      servicediscovery.RegistryLifecyclerConfig `yaml:"registry"`
 
 	HATrackerConfig HATrackerConfig `yaml:"ha_tracker,omitempty"`
 
-	MaxRecvMsgSize      int           `yaml:"max_recv_msg_size"`
-	RemoteTimeout       time.Duration `yaml:"remote_timeout,omitempty"`
-	ExtraQueryDelay     time.Duration `yaml:"extra_queue_delay,omitempty"`
-	LimiterReloadPeriod time.Duration `yaml:"limiter_reload_period,omitempty"`
+	MaxRecvMsgSize  int           `yaml:"max_recv_msg_size"`
+	RemoteTimeout   time.Duration `yaml:"remote_timeout,omitempty"`
+	ExtraQueryDelay time.Duration `yaml:"extra_queue_delay,omitempty"`
 
 	ShardByAllLabels bool `yaml:"shard_by_all_labels,omitempty"`
 
 	// for testing
-	ingesterClientFactory client.Factory
+	ingesterClientFactory client.Factory `yaml:"-"`
+	registryKVStore       kv.Client      `yaml:"-"`
 }
 
 // RegisterFlags adds the flags required to config this to the given FlagSet
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.BillingConfig.RegisterFlags(f)
 	cfg.PoolConfig.RegisterFlags(f)
 	cfg.HATrackerConfig.RegisterFlags(f)
+	cfg.Registry.RegisterFlagsWithPrefix("distributor.registry.", f)
 
 	f.BoolVar(&cfg.EnableBilling, "distributor.enable-billing", false, "Report number of ingested samples to billing system.")
 	f.IntVar(&cfg.MaxRecvMsgSize, "distributor.max-recv-msg-size", 100<<20, "remote_write API max receive message size (bytes).")
 	f.DurationVar(&cfg.RemoteTimeout, "distributor.remote-timeout", 2*time.Second, "Timeout for downstream ingesters.")
 	f.DurationVar(&cfg.ExtraQueryDelay, "distributor.extra-query-delay", 0, "Time to wait before sending more than the minimum successful query requests.")
-	f.DurationVar(&cfg.LimiterReloadPeriod, "distributor.limiter-reload-period", 5*time.Minute, "Period at which to reload user ingestion limits.")
 	f.BoolVar(&cfg.ShardByAllLabels, "distributor.shard-by-all-labels", false, "Distribute samples based on all labels, as opposed to solely by user and metric name.")
 }
 
 // New constructs a new Distributor
-func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Overrides, ring ring.ReadRing) (*Distributor, error) {
+func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Overrides, ring ring.ReadRing, internalDependency bool) (*Distributor, error) {
 	if cfg.ingesterClientFactory == nil {
```

Comment:
Can we rename it to `canJoinDistributorRing`, because if it follow the `ingestersRing` argument, then it seems reasonable that it joins the ingester ring and not a ring of its own.